lib/OrdinaryDiffEqDifferentiation/src/operators.jl

```diff
@@ -0,0 +1,301 @@
+            op.jvp_op(Jv, v, u, p, t)
+        else
+            copyto!(Jv, op.jvp_op(v, u, p, t))
+        end
+    end
+    return Jv
+end
+
+"""
+    StatefulJacobianOperator(jac_op::JacobianOperator, u, p, t)
+
+Wrapper over a [`JacobianOperator`](@ref) which stores the input `u`, `p` and `t`, and defines
+`mul!` and `*` for computing VJPs and JVPs.
+"""
+@concrete mutable struct StatefulJacobianOperator{M <: AbstractMode, T} <:
+                 AbstractJacobianOperator{T}
+    mode::M
+    jac_op <: JacobianOperator
+    u
+    p
+    t
+
+    function StatefulJacobianOperator(jac_op::JacobianOperator, u, p, t)
+        return new{
+            typeof(jac_op.mode), eltype(jac_op), typeof(jac_op), typeof(u), typeof(p), typeof(t)}(
+            jac_op.mode, jac_op, u, p, t)
+    end
+end
+
+Base.size(J::StatefulJacobianOperator) = size(J.jac_op)
+Base.size(J::StatefulJacobianOperator, d::Integer) = size(J.jac_op, d)
+
+for op in (:adjoint, :transpose)
+    @eval function Base.$(op)(operator::StatefulJacobianOperator)
+        return StatefulJacobianOperator($(op)(operator.jac_op), operator.u, operator.p, operator.t)
+    end
+end
+
+Base.:*(J::StatefulJacobianOperator, v::AbstractArray) = J.jac_op(v, J.u, J.p, J.t)
+Base.:*(J::StatefulJacobianOperator, v::Number) = J.jac_op(v, J.u, J.p, J.t)
+
+function LinearAlgebra.mul!(
+        Jv::AbstractArray, J::StatefulJacobianOperator, v::AbstractArray)
+    J.jac_op(Jv, v, J.u, J.p, J.t)
+    return Jv
+end
+
+
+
+# helper functions
+
+prepare_vjp(::Val{true}, args...; kwargs...) = nothing
+
+function prepare_vjp(
+        ::Val{false}, f::DiffEqBase.AbstractDiffEqFunction, u, p, t, fu; autodiff = nothing)
+    SciMLBase.has_vjp(f) && return f.vjp
+
+    autodiff isa AutoSparse ? autodiff = ADTypes.dense_ad(autodiff) : autodiff = autodiff
+
+    if isnothing(autodiff) && SciMLBase.has_jac(f)
+        if SciMLBase.isinplace(f)
+            jac_cache = similar(u, eltype(fu), length(fu), length(u))
+            return @closure (vJ, v, u, p, t) -> begin
+                f.jac(jac_cache, u, p, t)
+                LinearAlgebra.mul!(vec(vJ), jac_cache', vec(v))
+                return
+            end
+            return vjp_op
+        else
+            return @closure (v, u, p, t) -> reshape(f.jac(u, p, t)' * vec(v), size(u))
+        end
+    end
+
+    @assert autodiff!==nothing "`vjp_autodiff` must be provided if `f` doesn't have \
+                               analytic `vjp` or `jac`."
+
+    if SciMLBase.isinplace(f)
+        @assert DI.check_inplace(autodiff) "AD backend $(autodiff) doesn't support in-place problems."
+
+        fu_cache = copy(fu)
+
+        di_prep = DI.prepare_pullback(
+            f, fu_cache, autodiff, u, (fu,), DI.Constant(p), DI.Constant(t))
+        return @closure (vJ, v, u, p, t) -> begin
+            DI.pullback!(f, fu_cache, (reshape(vJ, size(u)),), di_prep, autodiff, u,
+                (reshape(v, size(fu_cache)),), DI.Constant(p), DI.Constant(t))
```

Comment:
Beware that the output of `reshape` may have a different type than the `fu` you prepared with, thus making preparation invalid

Reply:
wouldn't this just be a bug in `reshape`?